cli/src/api/api.ts

```diff
@@ -0,0 +1,73 @@
+import fetch from 'node-fetch';
+import chalk from 'chalk';
+import qs from 'query-string';
+import { getConfig } from '../config';
+
+const debug = process.env.API_DEBUG === 'true';
+
+export class ApiError extends Error {
+  constructor(public status: number, public body: {
+    error: string
+    invalidFields?: { path: string[], message: string, type: string }[]
+  }) {
+    super();
+  }
+}
+
+export async function api<T = any>(
+  { method, path, data, query } : {
+    method: 'get' |'post' | 'put' | 'delete'
+    path: string,
+    query?: Record<string, string | boolean | number | Date | undefined>,
+    data?: unknown
+  }
+): Promise<{ body: T } > {
+  const config = getConfig();
+
+  const url = `${config.apiUrl}${path}${query ? `?${qs.stringify(query)}` : ''}`
+
+  const headers = {
+    'Content-Type': 'application/json',
+    authorization: `token ${config.credentials}`,
+  }
+
+  if (debug) {
+    console.log(
+      chalk.cyan`[HTTP REQUEST]`,
+      chalk.dim(method?.toUpperCase()),
+      url,
+      chalk.dim(JSON.stringify({headers}))
+    );
+    if (data) {
+      console.log(chalk.dim(JSON.stringify(data)));
+    }
+  }
+  const response = await fetch(url, {
+    method,
+    headers,
+    body: data ? JSON.stringify(data) : undefined
+  });
+
+  const body = response.headers.get('Content-Type')?.startsWith('application/json') ?
+    await response.json() as T :
+    response.body as unknown as T;
+
+  if (debug) {
+    console.log(
+      chalk.cyan`[HTTP RESPONSE]`,
+      url,
+      chalk.cyan(response.status),
+      chalk.dim(JSON.stringify({ headers: response.headers }))
+    );
+    if (body) {
+      console.log(chalk.dim(JSON.stringify(body, null, 2)));
+    }
+  }
+
+  switch (response.status) {
+    case 200:
+      return { body };
+    default:
+      throw new ApiError(response.status, body as any);
```

Comment:
The fact that body can't be a stream here is not obvious from surrounding code.

I'd suggest to handle log streaming in a separate function.

Reply:
We just need a type checking here

I`d suggest to explicitly set expecting response type
```
const { data } = await axios.get('https://stream.example.com', {
    headers: {Authorization: `Bearer ${token}`, 
    responseType: 'stream'
});

// data is stream
```